add test-dds-adapter-depth test

maloel opened the pull request:
Adds basic streaming test for rs-dds-adapter.

Moved the `librs` library from dds-specific code to `rspy`, so other tests can use it.

Tracked on [LRS-1217]

OhadMeir commented:
> Moved the `librs` library from dds-specific code to `rspy`, so other tests can use it.

I am not sure `librs` is the appropriate name for this in `rspy` scope. Was named `librs` because it was a helper to test librealsense DDS capabilities (also testing `realdds` in same tests folder). Maybe incorporate it into `devices.py`?
What do you think?

OhadMeir reviewed (changes requested):
Thought about it, we have a `live/tools` unit test folder.  I think the adapter test should be there. Maybe under an internal `dds` folder

maloel commented:
> > Moved the `librs` library from dds-specific code to `rspy`, so other tests can use it.
> 
> I am not sure `librs` is the appropriate name for this in `rspy` scope. Was named `librs` because it was a helper to test librealsense DDS capabilities (also testing `realdds` in same tests folder). Maybe incorporate it into `devices.py`? What do you think?

It's librealsense, with everything that it has, plus some additions for unit-testing (but not for dds). Has nothing to do with dds.

maloel commented:
> Thought about it, we have a `live/tools` unit test folder. I think the adapter test should be there. Maybe under an internal `dds` folder

I don't agree: everything for dds should be under dds. The 'live' folder is not always right, and using it to group things does not always make sense. That's why we have `--live` and `--not-live` as switches for `run-unit-tests`. It was a good grouping at the time, but it's no longer needed.

If you insist, I'll change.

OhadMeir commented:
> > Thought about it, we have a `live/tools` unit test folder. I think the adapter test should be there. Maybe under an internal `dds` folder
> 
> I don't agree: everything for dds should be under dds. The 'live' folder is not always right, and using it to group things does not always make sense. That's why we have `--live` and `--not-live` as switches for `run-unit-tests`. It was a good grouping at the time, but it's no longer needed.
> 
> If you insist, I'll change.

This test should fail if no USB camera is attached. How is it marked to run only with live cameras now?

maloel commented:
> > > Thought about it, we have a `live/tools` unit test folder. I think the adapter test should be there. Maybe under an internal `dds` folder
> > 
> > 
> > I don't agree: everything for dds should be under dds. The 'live' folder is not always right, and using it to group things does not always make sense. That's why we have `--live` and `--not-live` as switches for `run-unit-tests`. It was a good grouping at the time, but it's no longer needed.
> > If you insist, I'll change.
> 
> This test should fail if no USB camera is attached. How is it marked to run only with live cameras now?

`--live` does not check the tag or folder - it checks whether it has a `test:device` directive in it.

maloel commented:
> `--live` does not check the tag or folder - it checks whether it has a `test:device` directive in it.

Not sure you saw this... I got no email...

OhadMeir commented:
OK if `--live` is looking for `device` directive, but I don't see that your test is running. It is skipped both for GHA and LibCI.

maloel commented:
> I don't see that your test is running

Good catch.
The reason is that this test is live and DDS at the same time, and we have no setup that runs both together...
We have to make LibCI compile DDS (possibly only on nightly?) @Nir-Az any preference?

maloel commented:
> 1 thing I am not sure from the code is if dds on will run only non live test or both live and not live?

I will make it so that, on nightly only, LibCI will compile DDS. This will be a configuration in Jenkins.
This will also cause the nightly to run all tests -- not just live -- i.e., including DDS -- and this test would then run.

maloel commented:
I had a lot of trouble - the adapter test failed in LibCI under Jenkins, and Dima could not help. It turns out that the `std::cin.ignore( ... )` code doesn't work under jenkins: `cin` reports it's at EOF and the adapter quits right away.

To get around this, I implemented proper Control-C handling in Windows/Linux and added it to rsutils.

You can see the run here:
https://rsjenkins.iil.intel.com/job/LRS_libci_pipeline/9859/

OhadMeir commented on third-party/rsutils/include/rsutils/concurrency/event.h:
```diff
@@ -0,0 +1,100 @@
+            std::unique_lock< std::mutex > lock( _m );
+            _is_set = true;
+        }
+        _cv.notify_all();
```
Shouldn' t `notify_all` also be under the lock?

In case thread A calls `set` and immediately after thread B calls `clear_and_set` we will want thread B to be blocked. Now we can have a case where `set` lock is released, thread B waits, then thread A continues and notify the cv.

maloel replied:
> Shouldn' t `notify_all` also be under the lock?

No, it [doesn't have to be](https://en.cppreference.com/w/cpp/thread/condition_variable#:~:text=can%20be%20done%20after%20releasing%20the%20lock).

If you want I'll move it in but it clearly says, and demonstrates in all the examples, that it does not need to be in the lock.

> In case thread A calls set and immediately after thread B calls `clear_and_wait`

If a notify is warranted, then a notify should happen. If you want nobody to clear it until all the notify work is done, then you need to make sure the notifications are part of the mutex, but there's no way to do that without actually calling a notification callback from within the lock.

Usually the more likely scenario is that, _on notification_, a thread would then say 'OK let me wait for the next one' so would do a `clear_and_wait`. There's no way to guarantee a notification-per-set on a simple event mechanism. If you want that, use a `signal` object.

OhadMeir replied:
I know it doesn't have to be locked.
Keeping it under the lock can help with a certain race scenario, were `set` and `clear_and_wait` are called (almost) simultaneously. It won't prevent the race itself, and I agree it is not the common use case.
Your call if to keep like this or change.

maloel replied:
I don't know if it's a problem: A set was called, so notifications were sent. Even if someone did a clear in-between.
I think it's good as-is unless we can identify a use-case where it's the fault of the mechanism.

OhadMeir commented on third-party/rsutils/include/rsutils/concurrency/event.h:
```diff
@@ -0,0 +1,100 @@
+    bool clear_and_wait( std::chrono::duration< Rep, Period > const & timeout )
+    {
+        std::unique_lock< std::mutex > lock( _m );
+        _is_set = false;
```
Why is this line needed? Will be overwritten in next line

maloel replied:
Because the next line waits. So while it's waiting, `is_set()` should return false.

OhadMeir commented on unit-tests/dds/adapter/test-depth.py:
```diff
@@ -0,0 +1,86 @@
+        test.info( 'start_time', start_time )
+        test.info( 'end_time', end_time )
+        test.info( 'n_frames', n_frames )
+        test.check_approx_abs( n_frames / (end_time-start_time), 29.5, 1 )
```
Validation FPS KPI is 5%, for 30Hz can deviate by up to 1.5Hz.
Already seen a failure with ~28.95Hz, so better update this

maloel replied:
It was 30 +/- 1, which is why it failed on 28.95 which is why I changed.
It's now 28.5-30.5 (I've seen it over 30, too). I think it should be good, but if not, what exactly would you like here? 28.5-31.5?

OhadMeir replied:
Validation tests for 28.5-31.5, but FPS over 30 is less likely.
Leave it 28.5-30.5, if we will see issues with too high FPS we can adjust the test.

OhadMeir commented:
> I will make it so that, on nightly only, LibCI will compile DDS. This will be a configuration in Jenkins. This will also cause the nightly to run all tests -- not just live -- i.e., including DDS -- and this test would then run.

Is this handled?

maloel commented:
> > I will make it so that, on nightly only, LibCI will compile DDS. This will be a configuration in Jenkins. This will also cause the nightly to run all tests -- not just live -- i.e., including DDS -- and this test would then run.
> 
> Is this handled?

As soon as the PR is in. It's a change in Jenkins.
